### PR TITLE
Tidy-ups in dvsim.py command line handling and wildcard substitution

### PR DESCRIPTION
**Pull request by rswarbrick**
These tidy-ups were spurred by issue #3322. The last patch "fixes" things for that (improving the error message from completely inscrutable to just slightly confusing). The other patches are assorted tidy-ups that I made while trying to trace my way through what was going on.

The big change to `subst_wildcards` was in a previous PR (#2458), which seems to have been completely ignored. I'm re-proposing it here, because it's a heck of a lot easier to follow the code, and it already supports the functionality that was added a couple of months later in commit `dba5d79` (I wish I'd noticed when this PR came up!)

### CODE REVIEW

**Comment by sriyerg** [comment]
@rswarbrick, this PR bundles commits that are totally unrelated to each other. If it is not too much to ask, would you please separate them into different PRs? Review (and merging into master) will go a lot quicker / smoother.

**Comment by imphil** [comment]
@sriyerg you can review the commits individually by going to the "Commits" tab on top of this PR. @rswarbrick has already rebased this work several times, it's much easier if it can go in in one PR, especially since you can still review on a patch-by-patch basis.

File: util/dvsim/dvsim.py [hunk 1]
```diff
@@ -114,6 +114,30 @@ def resolve_max_parallel(arg):
     return 16
 
 
+def read_verbosity(arg):
```
Comment (sriyerg) [inline util/dvsim/dvsim.py:117]: The original motivation for `dvsim` was to remain agnostic to the underlying technology used for running the simulations (UVM DV, Verilator or whatever). The 'generic' tool switches {n, l, m, h, d} would then transform to simulation-technology specific verbosity levels supplied in the HJson. I prefer to keep the codebase free from any assumptions (such as simulations are -only- run with UVM enabled), and instead maintain things like these in the 'user' data (sim cfg HJson files). I accept the fact that there is a lack of documentation explaining the design choices. 

I also agree that the mysterious error messages that get thrown due to totally unrelated setup issues should be fixed, but maybe in some other way.

**Comment by sriyerg** [comment]
> @sriyerg you can review the commits individually by going to the "Commits" tab on top of this PR. @rswarbrick has already rebased this work several times, it's much easier if it can go in in one PR, especially since you can still review on a patch-by-patch basis.

Commits 1 & 3 can be merged asap. 

Commit #2 needs some scrutiny since it makes fundamental changes to the existing code (IIUC it does not fix an actual bug, its a cosmetic change for improved readability). It affects not just DV, but FPV, lint and synthesis flows as well. I totally appreciate those changes, but we ought to be maximizing our focus on actual DV work at this point in time. We are already in agreement that non-critical patches will be reviewed with lower priority. 

I have provided my rationale for reverting commit #4.

File: util/dvsim/dvsim.py (hunk 1 above)
Comment (rswarbrick) [reply util/dvsim/dvsim.py:117]: Would you accept a patch that changed the magic names in the hjson files from `l`, `m` etc. to `verbosity_to_uvm_l`, `verbosity_to_uvm_m` etc?

**Comment by rswarbrick** [comment]
> Commit 2 needs some scrutiny since it makes fundamental changes to the existing code (IIUC it does not fix an actual bug, its a cosmetic change for improved readability). It affects not just DV, but FPV, lint and synthesis flows as well. I totally appreciate those changes, but we ought to be maximizing our focus on actual DV work at this point in time. We are already in agreement that non-critical patches will be reviewed with lower priority.

This is pretty frustrating. Commit 2 was pushed up in early June (as PR #2458). You didn't get around to reviewing it, but *did* re-implement a small part of it two months later (as PR #3085). If you'd reviewed it (it even comes with tests!) and committed it, you wouldn't have had to do the work in August. I'd really rather not shove this into the long grass to get ignored again.

Also, it does fix a bug. When hacking on hjson files, I managed to trigger an infinite loop in dvsim (because of a broken recursion in the wildcard expansions). This probably shouldn't be possible.

**Comment by rswarbrick** [comment]
I've split the `subst_wildcard` patch out of the PR (now in #3345). I've started tracking the "dvsim does something confusing if you give it a random hjson file" problem with issue #3344. I suspect there's more that we need to do to tidy that up, but this should be a start.

**Review by sriyerg — APPROVED** [review]

**Comment by sriyerg** [comment]
> I've split the `subst_wildcard` patch out of the PR (now in #3345). I've started tracking the "dvsim does something confusing if you give it a random hjson file" problem with issue #3344. I suspect there's more that we need to do to tidy that up, but this should be a start.

Thanks, I will take a look at it sometime next week.

**Pull request merged**